columnq/src/table/parquet.rs

```diff
@@ -1,78 +1,89 @@
 use std::io::Read;
 use std::sync::Arc;
 
+use crate::error::ColumnQError;
+use crate::table::{TableLoadOption, TableSource};
+
 use datafusion::arrow;
 use datafusion::arrow::datatypes::Schema;
 use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::datasource::parquet::ParquetTable;
+use datafusion::datasource::TableProvider;
 use datafusion::parquet::arrow::{ArrowReader, ParquetFileArrowReader};
 use datafusion::parquet::file::reader::SerializedFileReader;
 use datafusion::parquet::file::serialized_reader::SliceableCursor;
 
-use crate::error::ColumnQError;
-use crate::table::TableSource;
-
-pub async fn to_mem_table(
-    t: &TableSource,
-) -> Result<datafusion::datasource::MemTable, ColumnQError> {
-    // TODO: make batch size configurable
-    let batch_size = 1024;
-
-    let mut schema: Option<Schema> = None;
-
-    let partitions: Vec<Vec<RecordBatch>> =
-        partitions_from_table_source!(t, |mut r| -> Result<Vec<RecordBatch>, ColumnQError> {
-            // TODO: this is very inefficient, we are copying the parquet data in memory twice when
-            // it's being fetched from http store
-            let mut buffer = Vec::new();
-            r.read_to_end(&mut buffer).map_err(|_| {
-                ColumnQError::LoadParquet("failed to copy parquet data into memory".to_string())
-            })?;
-
-            let file_reader = SerializedFileReader::new(SliceableCursor::new(buffer))
-                .map_err(ColumnQError::parquet_file_reader)?;
-            let mut arrow_reader = ParquetFileArrowReader::new(Arc::new(file_reader));
-
-            let record_batch_reader = arrow_reader
-                .get_record_reader(batch_size)
-                .map_err(ColumnQError::parquet_record_reader)?;
-
-            let batch_schema = arrow_reader.get_schema().map_err(|_| {
-                ColumnQError::LoadParquet("failed to load schema from partition".to_string())
+pub async fn to_mem_table(t: &TableSource) -> Result<Arc<dyn TableProvider>, ColumnQError> {
+    let lazy = match &t.option {
+        Some(TableLoadOption::parquet { lazy }) => lazy.unwrap_or(false),
+        _ => false,
+    };
+
```

Comment:
would be cleaner to wrap `to_mem_table` with a wrapper function that matches on table option:

```rust
pub async fn to_datafusion_table() -> Result<Arc<dyn TableProvider>, ColumnQError> {
    match t.option {
        Some(TableLoadOption::parquet { lazy }) if lazy == Some(true) => Ok(Arc::new(ParquetTable::try_new(t.parsed_uri()?, 4).map_err(|err| {
            ColumnQError::LoadParquet(format!("failed to load parquet: '{}'", err.to_string()))
        })?)),
        _ => to_mem_table(t),
    }
}
```